pyads/get_bad_active_creatives.py

```diff
@@ -0,0 +1,177 @@
+from googleads import ad_manager
+import argparse
+import xlsxwriter
+
+
+def fetch_line_items(client):
+
+  active_line_items = {}
+  db_line_items = shelve.open('data/line_items.db')
+
+  line_item_service = client.GetService('LineItemService', version='v201805')
+
+  # Create a statement to select line items.
+  statement = ad_manager.StatementBuilder()
+
+  # Retrieve a small amount of line items at a time, paging
+  # through until all line items have been retrieved.
+  response = line_item_service.getLineItemsByStatement(statement.ToStatement())
+
+  for i in tqdm(range(int(int(response['totalResultSetSize']) / statement.limit))):
+    response = line_item_service.getLineItemsByStatement(statement.ToStatement(
+    ))
+    if 'results' in response and len(response['results']):
+      for line_item in response['results']:
+        db_line_items[str(line_item['id'])] = line_item
+        if line_item['endDateTime'] is not None:
+          date = line_item['endDateTime']['date']
+          if datetime(date['year'], date['month'], date['day']) > datetime.now():
+            active_line_items[line_item['id']] = line_item['name']
+        else:
+          active_line_items[line_item['id']] = line_item['name']
+      statement.offset += statement.limit
+    else:
+      break
+
+  db_line_items['active_line_items'] = active_line_items
+  db_line_items.close()
+
+  print("LINE ITEMS FETCHED: {}".format(len(active_line_items)))
+
+
+def fetch_associatiated_creatives(client):
+
+  creatives = {}
+  db_creatives = shelve.open('data/creatives.db')
+
+  lica_service = client.GetService(
+    'LineItemCreativeAssociationService', version='v201805')
+
+  # Get All creatives associated with active line items
+  print('GET CREATIVES FROM LINE ITEMS')
+
+  for line_item_id in tqdm(active_line_items):
+    statement = (ad_manager.StatementBuilder())
+
+    while True:
+      response = lica_service.getLineItemCreativeAssociationsByStatement(
+        statement.ToStatement())
+      if 'results' in response and len(response['results']):
+        for lica in response['results']:
+          if lica['status'] == "ACTIVE":
+            creatives[lica['creativeId']] = lica['lineItemId']
+            db_creatives[str(lica['creativeId'])] = lica['lineItemId']
+        statement.offset += statement.limit
+      else:
+        break
+
+  db_creatives['creatives'] = creatives
+  db_creatives.close()
+
+  print('Numer of associations fetched: {}'.format(len(creatives)))
+
+
+def fetch_creatives(client, creatives_to_scan):
+
+  db_creatives_to_scan = shelve.open('data/creatives_to_scan.db')
+
+  creative_service = client.GetService(
+    'CreativeService', version='v201805')
+
+  # Look for text in all creatives
+  print('SCANNING CREATIVES')
+
+  for creative_id in tqdm(creatives_to_scan):
+    statement = (ad_manager.StatementBuilder()
+                 .Where('creativeId = :creativeId')
+                 .WithBindVariable('creativeId', creative_id))
+
+    while True:
```

Comment:
nit: you're using the same structure of the loop in 3 places, you can abstract it to an utility method:
```
def some_fancy_name(statement, fetch_func):
  while True:
   response = fetch_func(statement.ToStatement())
   if 'results' in response and len(response['results']):
     for i in response['results']:
       yield i
    statement.offset += statement.limit
  else:
    break

# use it:
for creative in some_fancy_name(statement, creative_service.getCreativesByStatement):
  db_creatives_to_scan[str(creative['id'])] = creative
```